Fixing the line breaks for plaintext versions of the generated mails

vuchl opened the pull request:
- changed the templates
- added line break replacing

eladnava commented on index.js:
```diff
@@ -85,6 +85,12 @@ Mailgen.prototype.generatePlaintext = function (params) {
+    output = output.replace(/(?:\r\n|\r|\n)/g, '');
+    
+    // Replace htl break tags with linebreaks
+    output = output.replace(/(?:\<br\s*\/\>)/g, '\n');
```
This looks good, however, we want to maintain backward compatibility with custom themes that don't have `<br />` in their syntax to signal a line break. So let's simply wrap the logic you wrote on line 89 and 92 with an `if` that checks whether `this.cachedPlaintextTheme` contains at least one `<br />`.

eladnava commented on index.js:
```diff
@@ -85,6 +85,12 @@ Mailgen.prototype.generatePlaintext = function (params) {
+    // Strip all linebreaks from the rendered plaintext
+    output = output.replace(/(?:\r\n|\r|\n)/g, '');
+    
+    // Replace htl break tags with linebreaks
```
Tiny mistake in comment `htl` instead of `html`.

eladnava commented on themes/cerberus/index.txt:
```diff
@@ -1,31 +1,39 @@
 <% }) -%>
 <% } %>
-
+<% if (locals.intro) { %>
```
The additional if statement here is redundant. Take a look at this code:

```js
<% if (locals.intro) { %>
<% intro.forEach(function (introItem) { -%>
<%- introItem %><br />
<% }) -%>
<% } %>
<% if (locals.intro) { %>
<br />
<% } %>
```

You could refactor it as follows:
```js
<% if (locals.intro) { %>
<% intro.forEach(function (introItem) { -%>
<%- introItem %><br />
<% }) -%>
<br />
<% } %>
```

All I did was remove the following:
```js

<% } %>
<% if (locals.intro) { %>
```

eladnava replied:
Same goes for `locals.dictionary`, `locals.action` and `locals.outro`. Also applies to other themes.

vuchl commented:
https://github.com/eladnava/mailgen/pull/22/commits/af3ade3b3de8fa63b3800f84e0cf660cc205c779
I updated the comment, got backwards compatible line break checks and updated the templates as of your suggestions.

eladnava commented on index.js:
```diff
@@ -85,6 +85,19 @@ Mailgen.prototype.generatePlaintext = function (params) {
     var output = ejs.render(this.cachedPlaintextTheme, ejsParams);
 
+    // Definition of the <br /> tag as a regex pattern
+    var breakTag = /(?:\<br\s*\/\>)/g;
```
Now that I think of it, it would be nice to also support `<br>` tags which are valid in HTML as well. So let's add a `?` after the `\s*\/` in the regular expression. What do you think?

eladnava commented on themes/cerberus/index.txt:
```diff
@@ -1,31 +1,31 @@
+<%- actionItem.instructions %><br />
+<%- actionItem.button.link %><br />
 <% }) -%>
+<br />
```
I think this `<br />` should be moved inside the `action.foreach` loop so that the output appears like this:

```
To get started with Mailgen, please click here:
https://mailgen.js/confirm?s=d9729feb74992cc3482b350163a1a010

To get started with Mailgen, please click here:
https://mailgen.js/confirm?s=d9729feb74992cc3482b350163a1a010
```

And not like this:
```
To get started with Mailgen, please click here:
https://mailgen.js/confirm?s=d9729feb74992cc3482b350163a1a010
To get started with Mailgen, please click here:
https://mailgen.js/confirm?s=d9729feb74992cc3482b350163a1a010

```

eladnava commented:
@vuchl Perfect! 💯 

Well done. 😄

eladnava commented on index.js:
```diff
@@ -85,6 +85,19 @@ Mailgen.prototype.generatePlaintext = function (params) {
     // Render the plaintext theme with ejs, injecting the data accordingly
     var output = ejs.render(this.cachedPlaintextTheme, ejsParams);
 
+    // Definition of the <br /> tag as a regex pattern
```
Ooh, just thought of one more thing. We should move this entire logic to `Mailgen.prototype.cacheThemes` right after `this.cachedPlaintextTheme` because it would be better to cache this regex call instead of running it every time (which will be slow when people use Mailgen to generate 1000+ emails at a time). 👍

vuchl replied:
Good thinking!

eladnava replied:
@vuchl My bad, we actually shouldn't do this, as we need the `ejs` statements to be parsed first (so that the foreach statements are executed), so that the `<br />` statements get output for each item we are printing.

This is what is output now if we run the replace logic in `cacheThemes()`:

```
Hi John Appleseed,

Welcome to Mailgen!
We're very excited to have you on board.
To get started with Mailgen, please click here:
https://mailgen.js/confirm?s=d9729feb74992cc3482b350163a1a010

Need help, or have questions?
Just reply to this email, we'd love to help.
Yours truly,
Mailgen

&copy; 2016 <a href="https://mailgen.js/" target="_blank">Mailgen</a>. All rights reserved.
```

Can you please undo the last commit?

vuchl replied:
I `git rebase -i HEAD~2` and `git push origin +master`so the last commit should be gone.

eladnava commented:
Excellent, looks good to me! 💯 

Thanks for your contribution! 😄
